ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,575 @@
+  * Technically acceptable in a Ractor, but even ignoring that Ractors are optional, it's expensive to redefine all the
+    things ahead of time in every Ractor.
+* Refinements to prevent illegal calls.
+  * These are only scoped to the file/module/class where they are defined, not good enough for transitive checks we
+    need.
+  * Unacceptable performance hit to refine all we need to refine.
+* `TracePoint` to prevent illegal calls.
+  * Can be enabled on a specific thread which is nice to not affect other Ruby code.
+  * We would only check `:call` events probably, so not as bad of performance as some other uses of TracePoint.
+  * If running inside a Ractor, this only affects that Ractor which is nice.
+  * This approach is similar to what we did in .NET.
+  * We need to measure performance here.
+* `Ractor`s to isolate state and prevent some illegal calls.
+  * Ractors are experimental and upon first use give this warning: "warning: Ractor is experimental, and the behavior
+    may change in future versions of Ruby! Also there are many implementation issues."
+    * This will scare users.
+  * Users to be able to not use Ractors. The question becomes which is the default, knowing we can never change the
+    default even if Ractors do stabilize.
+* `Binding` to isolate state and prevent some illegal calls.
+  * We can create a binding and eval code within, but to do this transitively, we'd have to monkey patch `require` to
+    force reload which is not acceptable.
+  * This is effectively what we did with the Python sandbox and it is a big pain for users to understand when they want
+    to reload vs "pass through".
+
+So, after reviewing the options, only `TracePoint` and `Ractor` are viable in some ways, and are still subject to
+performance evaluation.
+
+### TracePoint
+
+We will create a `TracePoint` and enable it for a thread only when a workflow task is running on that thread. This
+should not have any performance impact outside of workflows. Currently it is believed we only need the `:call` event. It
+will do a quick lookup on a hash (or a hash of hashes) to check whether the call is disallowed and if it is, an
+exception will be raised.
+
+If a user needs to disable this for any reason (e.g. OTel libraries or advanced metrics/logging calls), they can run
+their code inside a block given to `Workflow::Unsafe.illegal_call_tracing_disabled`. There will also be
+`disable_workflow_tracing` worker option that will default to `false`.
+
+* ❓ Determine performance impact with this present vs not
+
+### Ractors
+
+We will offer a way to run workflows in a `Ractor`. This will provide us protection against state mutation. The downside
+is that not only are Ractors not stable and may have bugs, but they also give off a warning. A workflow instance will be
+in a Ractor and Core activations will be communicated.
+
+* ❓ Should we enable this by default, disable this by default, or somehow force the user to make a choice
```

Comment:
I don't think anyone cares about choosing between these two really. I think we need to figure out what has acceptable performance and correctness and do that. The warning / experimental status for Ractors seems like a nonstarter, though.

But, other than that, I can't really imagine why someone is coming along and deciding "Oh I really need to use Ractors and not TracePoint" for some reason. The only thing they care about is that the workflows run fast & errors get caught.

That is, of course, excepting some other weird interactions that aren't mentioned here or we don't know about.

Reply:
> The warning / experimental status for Ractors seems like a nonstarter, though.

I am not sure this is true. The Ractor gives so much benefit to ensuring safe determinism, it's basically a language-supported sandbox. And if we don't enable them by default (or force the user to choose whether to use them for now), we never can enable them by default. And they are quite decent for now for limited uses from my understanding.

> I can't really imagine why someone is coming along and deciding "Oh I really need to use Ractors and not TracePoint" for some reason

This section is purely for Ractors, I think we will turn on TracePoint by default.

It is a very tough call here on whether to enable Ractors by default, because Ractors are a perfect fit for this use case and we would support turning them off, and if we have a non-Ractor default now, we will never be able to change to the safer Ractor default if we want to later.

Reply:
> we never can enable them by default.

So why is that true?

I think the warning isn't so much the problem as the content of the warning that basically says "these are buggy". If that's not _actually_ true, then cool, I think we can be fine with it, but, we should know that for sure.

Reply:
> So why is that true?

Because you can't go from a less-restrictive default to a more-restrictive default without breaking people. Ractors prevent state sharing (e.g. a global is different in each Ractor). Also, things that are shared across Ractors must be marked as such via https://docs.ruby-lang.org/en/master/Ractor.html#method-c-make_shareable (which is an awesome feature and we already require this be done with payload and failure converters knowing they may be used inside sandbox). So turning on Ractors by default when they weren't before would break a lot of people that were, say, using their logger or OTel impl or other things people sometimes opt out of sandboxes for.

> I think the warning isn't so much the problem as the content of the warning that basically says "these are buggy". If that's not actually true, then cool, I think we can be fine with it, but, we should know that for sure.

We definitely have to test here, and they may be buggy for some advanced uses and maybe not our own. But even if that's the case, defaulting to experimental thing is rough, even though plenty of Ruby users do use Ractors today.

Reply:
Ah. I see. But arguably they already need to not be doing all that stuff for proper workflows anyway?

I think, if we find that we're not running into implementation bugs, it can be an OK default.

Reply:
> Ah. I see. But arguably they already need to not be doing all that stuff for proper workflows anyway?

Usually, yes, but some things people accept they need to cheat (e.g. interceptor sending exception to sentry when not replaying, or a logger that writes to disk but is not the logger we offer)

> I think, if we find that we're not running into implementation bugs, it can be an OK default.

I'm leaning there too. But it will be super loud that we use them and how not to.

Reply:
My vote is to force users to explicitly specify whether to use Ractors or not.

If we think we might want to change the default in the future, I think it's fine to require users to explicitly specify it now and remove the requirement later (so long as we're not asking them to do that for too many things). This enables us to migrate to the more restrictive default for future new users without breaking existing users.

We did something similar for Update IIRC.

Reply:
Hrmm, no SDK requires extra options for running workflows. I think if we choose Ractors we can change the default to be less restrictive if we ever had to (though I don't suspect we ever would).

> We did something similar for Update IIRC.

This is not comparable I don't think because we _know_ we will provide a default in the future but it is not yet implemented. Here, we don't know, so this isn't necessarily temporary. I think we might as well decide now, because in 5 years I don't think our decision would change.

Reply:
Proposal updated to Ractors by default with easy, well-documented way to opt out.

Reply:
Works for me.